[MRG+2] Rewrite setup.py files to handle cython dependencies

lesteve opened the pull request:
Fixes #7094 and maybe related issues if there are other. Closes #7708.
#### What does this implement/fix? Explain your changes.

Basically build_tools/cythonize.py doesn't take into account cython files dependencies. For example if you change _criterion.pxd the dependents .pyx don't get recompiled, potentially yielding to confusing segmentation faults.

The thing is that cython takes into acount correctly if you write your setup.py carefull (mostly by using Cython.Build.cythonize and switching sources between .c and .pyx as appropriate). Compared to #7708 which is a work-around this PR is trying to tackle the root problem.

Upsides:
- make in in is almost instant if you do not change cython files. If you do change cython files, only affected cython files are generated. As as developer this cuts down the time of compiling scikit-learn by a whole lot.
- dependency taken correctly into account. Try touching sklearn/tree/_criterion.pxd, and you'll see that all the sklearn/tree cython files gets recompiled.

Downsides:
- the recompilation is based on file timestamps, so **only on Travis** we lose the caching based on content hash but as I argued in #7094 the caching of .c generated files only saves 8% on individual build.

This is WIP because I haven't disabled caching on Travis yet. Just wanted to show the changes in setup.py files and get feed-back for now.

ping @raghavrv, @jnothman, @amueller.
- [x] pandas changes its sdist to cythonize all the files, maybe we need to do something like this depending how we release. I need to double-check this. Edit: sdist seems to work fine, pandas approach is a bit different where they have custom classes they pass to cmdclass in setup.
- [x] try with older versions of cython. pandas has a > 0.19.1 requirement, not sure why. Edit: Cython >= 0.23 is mentioned in the doc/developers/advanced_installation.rst and I checked it does compile fine with 0.23 and the tests run fine on my machine.

lesteve committed:
Rewriting of cythonization in setup.py

By using Cython.Build.cythonize and switching between .c and .pyx files
as appropriate cython dependencies are correctly taken into account.

raghavrv commented:
+10000 Thanks heaps for working on this...

lesteve commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,36 @@ def atlas_not_found(blas_info_):
+    is_cpp_filename = filename.endswith('.cpp')
+
+    # files in src are .c and .cpp files that are not cython-generated
+    if 'src/' in filename and (is_c_filename or is_cpp_filename):
```
This is the most brittle part of the PR I think. I don't know whether there is a better way of doing it, suggestions welcome.

lesteve commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,36 @@ def atlas_not_found(blas_info_):
+        sources = [get_cython_source(filename) for filename in sources]
+    config.add_extension(name, sources, **kwargs)
+
+    config.ext_modules = cythonize(config.ext_modules)
```
Hmmm I need to double-check this, maybe it should be inside an `if is_dev_version`

lesteve commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,36 @@ def atlas_not_found(blas_info_):
+            raise ValueError('Please install cython in order '
+                             'to build a scikit-learn development version')
+
+        sources = [get_cython_source(filename) for filename in sources]
```
This is the part that switches between using .c{,pp} files when building a release and the .pyx when building a dev version.

lesteve reviewed (commented):
A few inline comments to help the review

lesteve committed:
Use cythonize once on the root config rather than in each subpackage

raghavrv commented:
Just to note that sometimes there are inter module dependencies that needs to be recythonized too... For instance any change in the tree's cython sources needs a recythonizing of the gradient boosting... Does this PR handle that?

lesteve committed:
Fix for Windows

lesteve commented:
> Just to note that sometimes there are inter module dependencies that needs to be recythonized too... For instance any change in the tree's cython sources needs a recythonizing of the gradient boosting... Does this PR handle that?

It seems like it does:

```
make in
touch sklearn/tree/_criterion.pxd
make in
```

Excerpt of the output:

```
Compiling sklearn/ensemble/_gradient_boosting.pyx because it depends on ./sklearn/tree/_criterion.pxd.
Compiling sklearn/tree/_tree.pyx because it depends on ./sklearn/tree/_criterion.pxd.
Compiling sklearn/tree/_splitter.pyx because it depends on ./sklearn/tree/_criterion.pxd.
Compiling sklearn/tree/_criterion.pyx because it depends on sklearn/tree/_criterion.pxd.
Compiling sklearn/tree/_utils.pyx because it depends on ./sklearn/tree/_criterion.pxd.
[1/5] Cythonizing sklearn/tree/_criterion.pyx
[2/5] Cythonizing sklearn/tree/_splitter.pyx
[3/5] Cythonizing sklearn/ensemble/_gradient_boosting.pyx
[4/5] Cythonizing sklearn/tree/_tree.pyx
[5/5] Cythonizing sklearn/tree/_utils.pyx
```

raghavrv commented:
Amazing *_*

TomDLT commented:
This is nice !

lesteve commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,42 @@ def atlas_not_found(blas_info_):
+    is_cpp_filename = filename.endswith('.cpp')
+
+    # files in src are .c and .cpp files that are not cython-generated
+    if ('src' + os.sep) in filename and (is_c_filename or is_cpp_filename):
```
This is the most brittle part of the PR I think. I don't know whether there is a better way of doing it, suggestions welcome.

I guess one possible improvement would be to add an additional argument in add_cython_extension like cython_generated_sources so that we can discriminate more easily betwen the .c files coming from cython and the ones in the repo.

raghavrv replied:
This is fine I think. Optionally you could just hardcode a list with all the non cython source paths...

amueller replied:
How about checking if there exists a ``pyx`` file with the same name in the same folder?

raghavrv replied:
+1

lesteve replied:
That's a simple way of doing it indeed.

TomDLT replied:
Or we could add `.pyx` extension when it's a cython file, and add a function like:

```py
import os.path

def no_cythonize(extensions, **_ignore):
    for extension in extensions:
        sources = []
        for sfile in extension.sources:
            path, ext = os.path.splitext(sfile)
            if ext in ('.pyx', '.py'):
                if extension.language == 'c++':
                    ext = '.cpp'
                else:
                    ext = '.c'
                sfile = path + ext
            sources.append(sfile)
        extension.sources[:] = sources
    return extensions
```
(copy-paste from [cython doc](http://cython.readthedocs.io/en/latest/src/reference/compilation.html))

lesteve replied:
I think it would make more sense this way around indeed. The .pyx would be mentioned in the setup.py and you would do the switching to .c or .cpp only for releases. I have to say I didn't fully understand the no_cythonize function when I read the doc but it makes more sense now (in particular the naming is confusing because it's not really related to Cython.Build.cythonize).

lesteve commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,42 @@ def atlas_not_found(blas_info_):
+    is_dev_version = not os.path.exists(os.path.join(top_path, 'PKG-INFO'))
+
+    if is_dev_version:
+        sources = [get_cython_source(filename) for filename in sources]
```
This is the part that switches between using .c{,pp} files when building a release and the .pyx when building a dev version.

lesteve reviewed (commented):
Added a few of my comments that got folded in outdated diffs.

amueller commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,42 @@ def atlas_not_found(blas_info_):
+            raise ValueError('Please install cython in order '
+                             'to build a scikit-learn development version')
+
+        config.ext_modules = cythonize(config.ext_modules)
```
So that takes care of the caching?

lesteve replied:
From https://cython.readthedocs.io/en/latest/src/reference/compilation.html#compiling-with-distutils:

> The cythonize command also allows for multi-threaded compilation and dependency resolution. Recompilation will be skipped if the target file is up to date with its main source file and dependencies.

amueller reviewed (commented):
This looks pretty good if it works as expected. Also, I feel like this shouldn't live in scikit-learn alone. I feel like at least numpy, pandas and scikit-image need the same. Shouldn't we share a solution?

lesteve commented:
> This looks pretty good if it works as expected.

It does work as expected in the few tests I run like touching the _criterion.pxd and making sure that all the dependents are recompiled (see https://github.com/scikit-learn/scikit-learn/pull/7719#issuecomment-255387168). If you have more complicated scenarios that could go wrong I'd be happy to try them out. 

> Also, I feel like this shouldn't live in scikit-learn alone. I feel like at least numpy, pandas and scikit-image need the same. Shouldn't we share a solution?

Maybe we could have this in scikit-learn for some time so that it gets stress-tested a bit before consolidating across multiple projects?

amueller commented:
> Maybe we could have this in scikit-learn for some time so that it gets stress-tested a bit before consolidating across multiple projects?

Yes, but we could also try reaching out to others and see what their approaches and requirements are?

jnothman commented:
Why WIP?

raghavrv commented:
>  Remove build caching from Travis

+:100:

lesteve commented:
As I said in the issue description:

> This is WIP because I haven't disabled caching on Travis yet. Just wanted to show the changes in setup.py files and get feed-back for now.

No one complained about the Travis cache being useless as a side effect of this PR so I just pushed a commit that removes the build products from theTravis cache.

lesteve committed:
Remove caching from Travis

Cython dependencies are taken care of by Cython.Build.cythonize and
based on file timestamps, so .C and .so files will always be rebuild
from scratch on each build in Travis.

lesteve committed:
Specify .pyx in setup.files for cython generated extensions

More natural this way. Tweak the extensions to generate from .c and .cpp
files for a release.

GaelVaroquaux commented:
>     Also, I feel like this shouldn't live in scikit-learn alone. I feel like at
>     least numpy, pandas and scikit-image need the same. Shouldn't we share a
>     solution?

> Maybe we could have this in scikit-learn for some time so that it gets
> stress-tested a bit before consolidating across multiple projects?

+1. And then contribute it to Cython, IMHO.

ogrisel reviewed (approved):
LGTM.

lesteve committed:
COSMIT Remove commented out code

jnothman commented on sklearn/utils/sparsetools/setup.py:
```diff
@@ -1,21 +1,18 @@
 import numpy
 
 
+
```
Please remove added newline

lesteve replied:
Interestingly flake8_diff.sh is not able to spot these. Quickly looking at it this is because we do the diff without context (`--unified=0`) in order to avoid noise coming from lines that the PR hasn't touched.

jnothman commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,38 @@ def atlas_not_found(blas_info_):
+        try:
+            from Cython.Build import cythonize
+        except ImportError:
+            raise ValueError('Please install cython in order '
```
Given that we keep having issues related to cython versioning, I think we should be checking for or specifying version.

jnothman commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,38 @@ def atlas_not_found(blas_info_):
+        extension.sources = sources
+
+
+def tweak_extensions(top_path, config):
```
Can we call this something less opaque: `conditional_cythonize` or `cythonize_if_development`?

jnothman commented on sklearn/_build_utils/__init__.py:
```diff
@@ -33,3 +35,38 @@ def atlas_not_found(blas_info_):
     return cblas_libs, blas_info
+
+
+def tweak_extensions_to_build_from_c_and_cpp_files(extensions):
```
maybe just `rename_cython_extensions`?

jnothman reviewed (changes requested):
I've not tested that this works. Otherwise, it looks good to me...

lesteve committed:
Check cython version is greater than 0.23

lesteve committed:
COSMIT better names for functions

lesteve commented:
@jnothman I have tackled your comments.

jnothman commented on sklearn/_build_utils/__init__.py:
```diff
@@ -57,16 +59,24 @@ def tweak_extensions_to_build_from_c_and_cpp_files(extensions):
-        tweak_extensions_to_build_from_c_and_cpp_files(config.ext_modules)
+        build_from_c_and_cpp_files(config.ext_modules)
     else:
+        message = ('Please install cython with a version >= 0.23'
```
We should really pull this min version out as a constant.

lesteve replied:
Good idea, done.

jnothman commented:
Otherwise LGTM.

lesteve committed:
flake8 fix (imported module not at top of file)

lesteve committed:
Install cython 0.23 for Python 2.6

now that cython >= 0.23 requirement is enforced in setup.py

lesteve committed:
Use module constant for minimum required cython version

lesteve committed:
Fix Travis install.sh

No easy way to put comments inside multi-line command

jnothman commented:
You can merge after CI approves. thanks!

raghavrv commented:
Thank a lot for this @lesteve!

jnothman commented:
Is there a benefit to having this in 0.18.1?

On 2 November 2016 at 23:00, Raghav RV <notifications@github.com> wrote:

> Thank a lot for this @lesteve <https://github.com/lesteve>!
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7719#issuecomment-257845107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69hzUbKLEXZENHwRSUVnqSZEiufFks5q6HtfgaJpZM4KdIXJ>
> .
>

raghavrv commented:
Having this included in the backports will help avoid travis cache based error in the backport PR #7672 :D

lesteve commented:
> Thank a lot for this @lesteve!

@raghavrv your work-around PR motivated me to find a better solution, so you kind of initiated this ;-) !

TomDLT commented:
Really enjoying this change, compiling is very much faster !
